pkg/api/persistentvolumeclaim/util.go

```diff
@@ -29,34 +29,25 @@ const (
 
 // DropDisabledFields removes disabled fields from the pvc spec.
 // This should be called from PrepareForCreate/PrepareForUpdate for all resources containing a pvc spec.
-func DropDisabledFields(pvcSpec, oldPVCSpec *core.PersistentVolumeClaimSpec) {
-	// If this is an update, and the existing PVC had a data source,
-	// don't make any changes.
-	if dataSourceInUse(oldPVCSpec) {
-		return
-	}
-
+func DropDisabledFields(pvcSpec *core.PersistentVolumeClaimSpec) {
 	// Drop the contents of the dataSourceRef field if the AnyVolumeDataSource
 	// feature gate is disabled.
 	if !utilfeature.DefaultFeatureGate.Enabled(features.AnyVolumeDataSource) {
 		pvcSpec.DataSourceRef = nil
 	}
+}
 
-	// If the new dataSourceRef field is in use, don't make any changes,
-	// because usage of the new field implies acceptance of the semantics
-	// of KEP 1495.
-	if pvcSpec.DataSourceRef != nil {
-		return
-	}
-
-	// Drop invalid data sources for compatibility with old behavior.
-	// Specifically, if this is an update of a PVC with no data source, or
-	// a creation of a new PVC, and either the AnyVolumeDataSource feature
-	// gate is disabled or the dataSourceRef field is not filled in, then
-	// we will drop "invalid" data sources (anything other than a PVC or a
-	// VolumeSnapshot) from this request as if an empty PVC had been
-	// requested.
-	if !dataSourceIsEnabled(pvcSpec) {
+// EnforceDataSourceBackwardsCompatibility drops the data source field under certain conditions
+// to maintain backwards compatibility with old behavior.
+// See KEP 1495 for details.
+// Specifically, if this is an update of a PVC with no data source, or a creation of a new PVC,
+// and either the AnyVolumeDataSource feature gate is disabled or the dataSourceRef field is
```

Comment:
Nitty: I wouldn't mention the feature gate here - it's not actually checked and it is irrelevant.  The field is either in use or not.  The gate is handled elsewhere.

Reply:
As discussed in slack - this is the right place to tell the full story about why it is called on both create and update paths, and how it handles back-rev clients